haskellPackages.cereal: fix for pre-GHC8

Profpatsch opened the pull request:
See commit message.
Should fix git-annex.

Next step would be to move all packages that deal with this kind of issue from 7.10 overrides to common.
Next step would be to make cabal2nix recognize conditionals.

cc @peti

bennofs commented on pkgs/development/haskell-modules/configuration-lts.nix:
```diff
@@ -1783,7 +1783,6 @@ self: super: {
   "ceilometer-common" = dontDistribute super."ceilometer-common";
   "cellrenderer-cairo" = dontDistribute super."cellrenderer-cairo";
   "cerberus" = dontDistribute super."cerberus";
-  "cereal" = doDistribute super."cereal_0_5_2_0";
```
Why did you remove this override? https://www.stackage.org/lts-6.9 looks like it still has that cereal version

Profpatsch replied:
Oh, wasn’t aware that it was automatically generated. Adding it back in …

Profpatsch replied:
Hm, that makes it fail again.

I wonder if I should add the same rule for 5.2 or just keep 5.3, since it works.

bennofs replied:
I think you should add the same rule for 5.2 if 5.2 fails as well.

Profpatsch replied:
Done.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -2,7 +2,17 @@
+
+let
+  # apply an override to a package if it’s build with GHC <8.0
+  ifLT8 = override: drv:
```
This function should be in `lib.nix` and it should have a more descriptive name than `ifLT8`.

Profpatsch replied:
I didn’t want to expose that (yet), since it’s not completely clear if it’s a helpful library function (at least for me).

peti replied:
Yeah, but it's confusing for readers of the file to define a "local helper function" some 1000 lines of code apart from the place where it's actually used.

peti commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -947,8 +957,11 @@ self: super: {
+  # cereal must have `fail` in pre-ghc-8.0.x versions
+  # also tests require bytestring>=0.10.8.1
+  cereal = ifLT8 (p: dontCheck (addBuildDepend p self.fail)) super.cereal;
+  cereal_0_5_2_0 = ifLT8 (p: dontCheck (addBuildDepend p self.fail)) super.cereal_0_5_2_0;
```
What I don't like about this approach is that we have a file `configuration-ghc-7.10.x.nix` that says "overrides applied for GHC 7.10.x are in here!", but in fact they are not. There is something to be said for duplicating those overrides in the various compiler configs, even if it is more verbose.

Profpatsch replied:
Would it make sense to have a new overwrite layer for these kind of overwrites? At least until we can generate it with cabal2nix? Something like `configuration-old-compilers.nix` maybe.

peti replied:
I feel like the current solution -- copy-pasting the override into those config files that need it -- works sufficiently well and I don't see much of an incentive to add another config file right now. Besides, it's not uncommon for those overrides to require even more extra build inputs the older the compiler gets, i.e. it's not always easy to define one override that works everywhere.

Profpatsch replied:
> Besides, it's not uncommon for those overrides to require even more extra build inputs the older the compiler gets

In that case the function can be applied multiple times with different compiler versions.
So should I copy and paste that into every 7.x configuration file?

peti replied:
> So should I copy and paste that into every 7.x configuration file?

I would prefer that.

Profpatsch committed:
haskellPackages.cereal: fix for pre-GHC 8.0

Profpatsch committed:
haskellPackages.Glob: remove superfluous override

The override is already in the 7.10.x file, and we’d like to explicitely
override for every version that needs it.

Profpatsch commented:
@peti I copied it to 7.10, so it should now work for lts as well.
